Open-Agriculture/AgIsoStack-plus-plus
# Add support for Arduino, Teensy 4 and 4.1

## What's new

Added lots of changes to allow the stack to build without threading under the Arduino IDE. Added CAN plugin for Teensy hardware using the FlexCAN T4 api.

Changes include:

* Added single threaded CANHardwareInterface option. Changed dynamic casts to static casts to completely eliminate rtti, which shouldn't be an issue since the type can be inferred in those cases where we were using dynamic_pointer_cast.
* Added preprocessor checks to optionally compile out all threading. Added a script to automatically generate an Arduino library from the content of the repo.
* Added Arduino example files.
* Added a python script to automatically generate an Arduino library based on the content of the repo.

Tested with Teensy 4.1 hardware and SN65HVD230 transceiver.

### Follow-up tasks

- [x] Create new repository to store the latest-generated Arduino library version of the stack
- [ ] Submit library to the Arduino [library registry](https://github.com/arduino/library-registry)
- [ ] Add CI pipeline for building with Arduino
- [ ] Investigate adding pipeline that can automatically generate the Arduino library and commit it to the other repo, plus open a PR
- [ ] Create documentation for using the library on Arduino

### How to use/test this right now

Until all the tasks above are done, here's how to try this on Teensy 4.1 on Windows.

1. Install [Arduino IDE](https://www.arduino.cc/en/software)
2. Add Teensy board support to Arduino IDE by following [these instructions](https://www.pjrc.com/teensy/td_download.html)
3. Open the [example file](https://github.com/Open-Agriculture/AgIsoStack-plus-plus/blob/arduino/examples/arduino_example/arduino_example.ino) in your Arduino IDE.
4. Select your board as shown:
![image](https://github.com/Open-Agriculture/AgIsoStack-plus-plus/assets/10929341/59c5dce3-905f-4590-9f9a-7559bad09ae9)
5. Ensure optimizations are set to "Smallest Code"
![image](https://github.com/Open-Agriculture/AgIsoStack-plus-plus/assets/10929341/13c32af4-e37b-456b-89f5-15a484fb24a4)
6. There's some kind of bug in the Teensy's core, where if you are not building with the default optimization settings, the board fails to start up. Some [attempts](https://github.com/PaulStoffregen/cores/pull/673) appear to have been made to fix this but it remains an issue for me at least. To ensure your board starts up, add `#pragma GCC optimize ("O2")` to the teensy's `startup.c` file which is likely located at the following location: `C:\Users\<your user name>\AppData\Local\Arduino15\packages\teensy\hardware\avr\1.58.1\cores\teensy4`
![image](https://github.com/Open-Agriculture/AgIsoStack-plus-plus/assets/10929341/ec7159db-b54c-4a34-ae0d-b73575990775)
7. Run the `generateArduinoLibrary.py` script with Python in the root of the repo. I usually do this in IDLE because I'm a Python neophyte basically. I know C++ things, what can I say.
![image](https://github.com/Open-Agriculture/AgIsoStack-plus-plus/assets/10929341/3b73a303-1529-41d0-b2b1-1715a3850213)
9. Copy the newly created `arduino_library` folder that the script creates to your Arduino `libraries` directory. You may need to create this directory if it doesn't exist, but it's usually located at `C:\Users\<your user name>\Documents\Arduino\libraries`. Maybe rename the `arduino_library` folder to AgIsoStack or something if you want.
10. Build the sketch, and upload it to your board 🙂

You'll also need a CAN transceiver connected to your Teensy's CAN1 pins.

Closes #264
